control-plane-operator/main.go

```diff
@@ -237,13 +225,31 @@ func NewStartCommand() *cobra.Command {
 			os.Exit(1)
 		}
 
-		lookupOperatorImage := func(deployments appsv1client.DeploymentInterface, name, userSpecifiedImage string) (string, error) {
+		// For now, since the hosted cluster config operator is treated like any other
+		// release payload component but isn't actually part of a release payload,
+		// enable the user to specify an image directly as a flag, and otherwise
+		// try and detect the control plane operator's image to use instead.
+		lookupOperatorImage := func(userSpecifiedImage string) (string, error) {
 			if len(userSpecifiedImage) > 0 {
 				setupLog.Info("using image from arguments", "image", userSpecifiedImage)
 				return userSpecifiedImage, nil
 			}
-			deployment, err := deployments.Get(context.TODO(), name, metav1.GetOptions{})
-			if err != nil {
+			if podName := os.Getenv("POD_NAME"); podName != "" {
+				me := &corev1.Pod{ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: podName}}
+				if err := mgr.GetAPIReader().Get(ctx, crclient.ObjectKeyFromObject(me), me); err != nil {
+					return "", fmt.Errorf("failed to get operator pod %s: %w", crclient.ObjectKeyFromObject(me), err)
+				}
+				// Use the container status to make sure we get the sha256 reference rather than a potentially
+				// floating tag.
+				for _, container := range me.Status.ContainerStatuses {
+					// TODO: could use downward API for this too, overkill?
+					if container.Name == "control-plane-operator" {
+						return container.ImageID, nil
+					}
+				}
+			}
+			deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: deploymentName}}
+			if err := mgr.GetAPIReader().Get(ctx, crclient.ObjectKeyFromObject(deployment), deployment); err != nil {
```

Comment:
Does this work before the manager is started?

Reply:
Yes, the manager not being started is why it using the APIReader rather than the normal client, which wouldn't work (and also to avoid creating a cache for pods)